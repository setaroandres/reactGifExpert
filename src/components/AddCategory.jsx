import { useState } from 'react';
import propTypes from 'prop-types';


export const AddCategory = ({onNewCategory}) => { //Aca recibimos lo que enviamos desde le padre (props)

	//Cada componente puede tener sus propios hooks
	const [inputValue, setInputValue] = useState('');

	const onInputChange = ({target}) => {
		setInputValue(target.value);
	}

	const onSubmit = (event) => {
		event.preventDefault();//Para evitar el refresh por defecto
		//Aca tenemos que hacer el intercambio de informacion con el componenet padre para mandarle el valor del input. Aca llamamos a la function que comunica con el padre
		//y le mandamos los nuevos valores de las categorias para que se actulicen en setCategories
		if (inputValue.trim().length <= 1) return;
		onNewCategory(inputValue.trim());//Aca emitimos el valor del input value, es como el emit de Angular

		setInputValue('');
	}

	return (
		<form onSubmit={(event) => onSubmit(event)} aria-label='form'> 
			<input 
				type="text"
				placeholder="Buscar Gifs"
				value={inputValue}
				onChange={onInputChange} //Lo podemos poner asi pq siempre mandamos el event, asi nos obviamos pasarlo por el param
			/>
		</form>
	)
}

AddCategory.propTypes = {
	onNewCategory: propTypes.func.isRequired
}