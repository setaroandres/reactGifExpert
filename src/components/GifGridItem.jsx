import propTypes from 'prop-types';

export const GifGridItem = ({title, url}) => {
  return (
            <div className="card">
                <img src={url} alt={title} />
                <p>{title}</p>
            </div>
  )
}

GifGridItem.propTypes = {
  title: propTypes.string.isRequired, 
  url  : propTypes.string.isRequired,
}