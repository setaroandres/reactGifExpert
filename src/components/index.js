//Aca podemos poner todos los exports ara tenerlos todos aca

export * from './AddCategory';
export * from './GifGrid';
export * from './GifGridItem';