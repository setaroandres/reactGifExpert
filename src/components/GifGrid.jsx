import propTypes from 'prop-types';
//La fcn de fetch de los gifs esta en helpers

import { useEffect, useState } from "react";
import { GifGridItem } from "./GifGridItem";
import { useFetchGifs } from "../hooks/useFetchGifs";

export const GifGrid = ({category}) => {

	//Aca creamos el custom hook para poenr toda la logica del getGifs. Hacemos desestructuracion de obj para poder usar y de volver lo que queramos del hook.
	const {images, isLoading} = useFetchGifs(category);

	//console.log({images, isLoading});

	return (
			<>
				<h3>{category}</h3>
				{
					isLoading && (<h2>Cargando...</h2>)//el && es un if corto....is isloading = true en este caso
				}
				
				<div className="card-grid">
					{
						images.map((image) => 
							(
								<GifGridItem 
									key={image.id}
									{...image}//Todas las prop que vienen en el image las estamos esparciendo y las mandamos al hijo
								/>
							)
						)
					}
				</div>
			</>
	)
}

GifGrid.prototype = {
	category: propTypes.string.isRequired
}