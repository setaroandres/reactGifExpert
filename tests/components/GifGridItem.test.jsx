import { render, screen } from "@testing-library/react";
import { GifGridItem } from "../../src/components/GifGridItem";

describe('Pruebas en GifGridItem', () => { 

    const title = 'The Simpsons';
    const url = 'http://thesimps.com/bart.gif'

    test('Debe hacer match con el snapshot', () => { 
        const {container} = render(<GifGridItem title={title} url={url}/>);
        expect(container).toMatchSnapshot();
    });

     test('Debe mostrar la imagen con el url y el alt indicado', () => { 
        //siempre tenemos que tomar el sujeto de pruebas y renderizarlo
        render(<GifGridItem title={title} url={url}/>);
        //screen.debug();
        // expect(screen.getByRole('img').src).toBe(url);
        // expect(screen.getByRole('img').alt).toBe(title);
        const {src, alt} = screen.getByRole('img');
        expect(src).toBe(url);
        expect(alt).toBe(title);
    });

    test('Debe de mostrar el titulo en el componente', () => { 
        //siempre tenemos que tomar el sujeto de pruebas y renderizarlo
        render(<GifGridItem title={title} url={url}/>);
        expect(screen.getByText(title)).toBeTruthy();
    });
});