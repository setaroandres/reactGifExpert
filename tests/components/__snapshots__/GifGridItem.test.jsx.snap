// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifGridItem Debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="The Simpsons"
      src="http://thesimps.com/bart.gif"
    />
    <p>
      The Simpsons
    </p>
  </div>
</div>
`;
